guides/type_definitions/objects.md

```diff
@@ -0,0 +1,282 @@
+
+```ruby
+field :top_score, Integer, null: false
+```
+
+The default behavior is to look for a `#top_score` method, or lookup a hash key, `:top_score`.
+
+You can override the method name with the `method:` keyword, or override the hash key with the `hash_key:` keyword, for example:
+
+```ruby
+# Use the `#best_score` method to resolve this field
+field :top_score, Integer, null: false, method: :best_score
+# Lookup `hash["allPlayers"]` to resolve this field
+field :players, [User], null: false, hash_key: "allPlayers"
+```
+
+If you don't want to delegate to the underlying object, you can define a method for each field:
+
+```ruby
+# Use the custom method below to resolve this field
+field :total_games_played, Integer, null: false
+
+def total_games_played
+  @object.games.count
+end
+```
+
+Inside the method, you can access some instance variables:
+
+- `@context` is the query context (formerly `ctx` to resolve functions)
+- `@object` is the underlying application object (formerly `obj` to resolve functions)
+
+Additionally, when you define arguments (see below), they're passed to the method definition, for example:
+
+```ruby
+# Call the custom method with incoming arguments
+field :current_winning_streak, Integer, null: false do
+  argument :include_ties, Boolean, required: false, default_value: false
+end
+
+def current_winning_streak(include_ties:)
+  # Business logic goes here
+end
+```
+
+### Field Arguments
+
+_Arguments_ allow fields to take input to their resolution. For example:
+
+- A `search()` field may take a `term:` argument, which is the query to use for searching, eg `search(term: "GraphQL")`
+- A `user()` field may take an `id:` argument, which specifies which user to find, eg `user(id: 1)`
+- An `attachments()` field may take a `type:` argument, which filters the result by file type, eg `attachments(type: PHOTO)`
+
+Arguments can be expressed in the SDL:
+
+```ruby
+type User {
+  # This user's transaction history, after `since` if present.
+  transactions(since: DateTime): [Transaction!]!
+}
+```
+
+Arguments are _typed_, so each argument takes a certain kind of data. Only a few types are valid inputs:
+
+- __Scalars__, such as `String`, `Integer`, `Float`, `Boolean`, `ID`, or custom scalar types
+- __Enums__, defined by your application
+- __Input objects__, defined by your application
+- __Lists__ of any of the above type
+
+To add arguments to fields, use the `argument(...)` method, inside a block:
+
+```ruby
+field :transactions, [Types::Transaction], null: false do
+  argument :since, Types::DateTime, required: false
+end
+```
+
+If an argument has `required: true`, then all queries to the field _must_ provide a value for that argument. `required: false` means that the argument is optional.
+
+Arguments can also accept a description and a `default_value:`, for example:
+
+```ruby
+field :transactions, [Types::Transaction], null: false do
+  # Description is added after the type name
```

Comment:
Not entirely relevant to this PR, but I didn't realize this was a positional arg. Wouldn't a kwarg be more consistent with how `field` does it?

Reply:
/shrug `field` does _both_ positional and kwarg. Fine with me to make `argument` also do both, but I haven't done it yet.

Reply:
addressed in #1280